### ISSUE TITLE
Memory Leak?

### ISSUE THREAD

**Issue by lhk** [croian/mouseless-issues#16]
### mouseless version

0.2.0

### macOS version

15.1.1

### What happened?

Memory consumption keeps slowly climbing up:
<img width="630" alt="image" src="https://github.com/user-attachments/assets/3533aeb5-8f8b-4e1b-9211-46a5a01bbdcf" />

After a few hours of mouseless idling in the background, it reaches 750MB.
When quitting and restarting it's at 150MB

### Steps to Reproduce

Start it, let it run and monitor memory consumption.

**Comment by Zer0x00** [comment]
3 GB RAM usage for me after one day of work

**Comment by croian** [comment]
yikes, my apologies to everyone for this issue.  if I can quickly diagnose and implement a fix for the first patch release later today, will do so.  otherwise, it will be my highest priority for the next release.

**Comment by lhk** [comment]
@croian do you actually use the tool during daily coding, do you have this switched on?

sorry, I am really excited for this. But learning to use such a tool efficiently is a bit of a leap of faith. Takes some time investment before I'd expect this to be comfortable. And I'd feel more confident in actually giving mouseless a shot if you can vouch for this as part of your own development routine. But with gigabytes of RAM being leaked after just one day, it doesn't seem like this can be used routinely.

**Comment by croian** [comment]
@lhk i do use it daily, but the irony is that *while developing it* there are difficulties in having it running all the time from its built/standalone version, as it can conflict and get confusing when running alongside the dev version -- so i usually rely on the dev version, which i am continually restarting throughout the day due to making code changes, etc.

i'll have more content soon that demonstrate when/how its useful to me, and I hope it can be useful to you as well.  either way, I certainly appreciate your feedback and the bug report, so thank you!

**Comment by croian** [comment]
also, regarding time investment, at least one youtube commenter said they were getting the hang of it after just five minutes, which I'm heartened by!  i think it will become intuitive pretty quickly for already keyboard-centric users

**Comment by croian** [comment]
quick update before i wrap things up and start putting together the patch release -- I've managed to *significantly* alleviate the issue, but some work remains.  users should probably see memory grow at 1-5% the prior rate, and with some work put in tomorrow/this weekend I should be able to figure out the rest.

thank you everyone for your patience while I iron the kinks out!

**Comment by croian** [comment]
partial fix has been released in v0.2.1

**Comment by LouisLafontaine** [comment]
> also, regarding time investment, at least one youtube commenter said they were getting the hang of it after just five minutes, which I'm heartened by! i think it will become intuitive pretty quickly for already keyboard-centric users

Yeah, I have been using for the last 5 mins, and I've already got a hang of it too. Changed the overlay to be more readable. I think it's pretty easy to use. Only thing I wish for is that you could bind two keys to an action, like (AltRight or Escape) to exit

**Comment by croian** [comment]
@LouisLafontaine that's really great to hear!  i will be adding multiple keybindings per action in an upcoming release, will definitely be useful

**Comment by croian** [comment]
Update:  I've observed that in addition to the 1-5% remaining memory leak, there are spikes (typically 170mb, sometimes more) in memory usage when displays are added/removed/resized.  These should be minimized in duration (if not eliminated entirely) in v0.2.2.

Additionally, there is an optimization for overlay resource usage that I have been planning, which will be a bit labor-intensive, but I think will help improve baseline memory usage overall (including reduced magnitude of any spikes), perhaps significantly.  Will probably slot this into the release after v0.2.2.

**Comment by lhk** [comment]
>  -- so i usually rely on the dev version, which i am continually restarting throughout the day due to making code changes, etc.

Ah that makes a lot of sense.
I've taken the plunge and paid for the app. Really love the idea. And it's impressive how fast you're iterating on bugfixes. Once this is a bit more stable I'll recommend it vocally to friends :)

**Comment by ognistik** [comment]
Just wanted to report that in version 0.2.1 I'm consistently seeing Mouseless use around 500 to 600 megabytes of RAM in the background when I'm not actively using it. Of course it goes higher when I do use it. I really hope this will still be considered for further improvements.

**Comment by croian** [comment]
@ognistik thank you for the info.  that's in the range i've seen as well, though on v0.2.2-1 i'm seeing ~245mb right now -- you may notice some improvement by trying one of the [preview releases](https://github.com/croian/mouseless-issues/releases)

this is *definitely* still a priority -- i think the big difference will be made by the optimization I mentioned [here](https://github.com/croian/mouseless-issues/issues/16#issuecomment-2543443805).  I aim to get this in ASAP, within 2 weeks at the very latest.

**Comment by croian** [comment]
UPDATE:  I made the big optimization I've been planning and solved the remaining memory leak, and it appears to have made a big difference.  These changes will be in the v0.3.0 release.

The results:

- Baseline real memory used is reduced by up to 40% (see Real Mem column in activity monitor or RES column in htop)
  -  I'm seeing a baseline of 175 to 190MB (a bit higher at startup, but stabilizes after one or two overlay usages) with 1 monitor, about 205MB with 2 monitors
    - This is vs prior baseline of 300MB+
- Shared memory usage is reduced by up to 95%
- Memory spikes further reduced when monitors added/removed/changed
- Real Mem usage should remain lower and more stable overall

**Note**: the `Memory` column typically shows a higher value (stabilizes to about 460MB) than the `Real Mem` column and may show spikes when the overlay is used and a click is made, but the `Real Mem` (the important one), `Private Mem`, and `Shared Mem` columns stay stable during usage.
  - This is likely the system allocating a memory block but not physically using it, so shouldn't actually have much impact on system performance (note the `Purgable Mem` column shows ~290MB, so a lot of that allocated memory is reclaimable by the OS)
  
I still have some optimizing to do when configuration changes are made, and further optimizations may be made as I work on cross-platform versions in the coming months -- part of that process might entail re-writing some of the code at a lower level across all platforms.

![mem_usage](https://github.com/user-attachments/assets/95bc9e41-ccba-4b21-878f-961e4aa5c76f)

**Comment by incanus** [comment]
Is this potentially still an issue in 0.4.0-preview.2? I see high memory usage frequently, actually the top process when I have no GUI apps running after `WindowServer`. As an example, I have rebooted several days ago and haven't yet used the app (I go periods of mouse vs. keyboard) and it's currently using 783.7MB. Now upon restarting it the utilization is 72.1MB even after several uses. 

Update: I am also noticing a rogue `about:` process which belongs to Mouseless and is using 121.2MB itself. It's entitled `about:` in the process list but looks like this: 

<img width="1110" alt="Image" src="https://github.com/user-attachments/assets/45091f19-f056-4664-b94f-c01d8fe2daff" />

**Comment by cymian** [comment]
Thank you for the report @incanus, I'll look into this in the week or two, as I work toward the stable 0.4 release.
